agent/taskresource/volume/volume_test.go

```diff
@@ -0,0 +1,142 @@
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package taskresource
+
+import (
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/engine"
+	docker "github.com/fsouza/go-dockerclient"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestCreateSuccess(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	mockClient := engine.NewMockDockerClient(ctrl)
+
+	name := "volumeName"
+	mountPoint := "some/mount/point"
+	driver := "driver"
+	driverOptions :=  map[string]string{
+		"opt1": "val1",
+		"opt2": "val2",
+	}
+
+	mockClient.EXPECT().CreateVolume(name, driver, driverOptions, nil, engine.CreateVolumeTimeout).Return(
+		engine.VolumeResponse{
+			DockerVolume: &docker.Volume{Name: name, Driver: driver, Mountpoint: mountPoint, Labels: nil},
+			Error: nil,
+	})
+
+	volume := NewVolumeResource(name, driver, driverOptions, nil, mockClient)
+	err := volume.Create()
+	assert.NoError(t, err)
+	assert.Equal(t, mountPoint, volume.Mountpoint)
+}
+
+func TestCreateError(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	mockClient := engine.NewMockDockerClient(ctrl)
+
+	name := "volumeName"
+	driver := "driver"
+	labels :=  map[string]string{
+		"label1": "val1",
+		"label2": "val2",
+	}
+
+	mockClient.EXPECT().CreateVolume(name, driver, nil, labels, engine.CreateVolumeTimeout).Return(
+		engine.VolumeResponse{
+			DockerVolume: nil,
+			Error: errors.New("some error"),
+	})
+
+	volume := NewVolumeResource(name, driver, nil, labels, mockClient)
+	err := volume.Create()
+	assert.NotNil(t, err)
+}
+
+func TestCleanupSuccess(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	mockClient := engine.NewMockDockerClient(ctrl)
+
+	name := "volumeName"
+	driver := "driver"
+
+	mockClient.EXPECT().RemoveVolume(name, engine.RemoveVolumeTimeout).Return(nil)
+
+	volume := NewVolumeResource(name, driver, nil, nil, mockClient)
+	err := volume.Cleanup()
+	assert.NoError(t, err)
+}
+
+func TestCleanupError(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	mockClient := engine.NewMockDockerClient(ctrl)
+
+	name := "volumeName"
+	driver := "driver"
+
+	mockClient.EXPECT().RemoveVolume(name, engine.RemoveVolumeTimeout).Return(errors.New("some error"))
+
+	volume := NewVolumeResource(name, driver, nil, nil, mockClient)
+	err := volume.Cleanup()
+	assert.NotNil(t, err)
+}
+
+func TestMarshall(t *testing.T) {
+	volumeStr := "{\"Name\":\"volumeName\",\"MountPoint\":\"\",\"Driver\":\"driver\",\"DriverOpts\":{},\"Labels\":{}," +
```

Comment:
You can use '\`' instead of `\"`

Reply:
I tried that for multi-line string, but those keep whitespaces between lines. I didn't want to add more logic to clean up whitespaces.